src/coreComponents/physicsSolvers/multiphysics/CoupledReservoirAndWellsBase.cpp

```diff
@@ -0,0 +1,291 @@
+
+  // Copy the original nonzeros
+  for( localIndex localRow = 0; localRow < patternDiag.numRows(); ++localRow )
+  {
+    globalIndex const * cols = patternDiag.getColumns( localRow ).dataIfContiguous();
+    pattern.insertNonZeros( localRow, cols, cols + patternDiag.numNonZeros( localRow ) );
+  }
+
+  // Add the nonzeros from coupling
+  addCouplingSparsityPattern( domain, dofManager, pattern.toView() );
+
+  // Finally, steal the pattern into a CRS matrix
+  localMatrix.assimilate< parallelDevicePolicy<> >( std::move( pattern ) );
+  localMatrix.setName( this->getName() + "/localMatrix" );
+
+  rhs.setName( this->getName() + "/rhs" );
+  rhs.create( dofManager.numLocalDofs(), MPI_COMM_GEOSX );
+
+  solution.setName( this->getName() + "/solution" );
+  solution.create( dofManager.numLocalDofs(), MPI_COMM_GEOSX );
+}
+
+template< typename RESERVOIR_SOLVER, typename WELL_SOLVER >
+void CoupledReservoirAndWellsBase< RESERVOIR_SOLVER, WELL_SOLVER >::
+assembleSystem( real64 const time_n,
+                real64 const dt,
+                DomainPartition & domain,
+                DofManager const & dofManager,
+                CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                arrayView1d< real64 > const & localRhs )
+{
+  // assemble J_RR (excluding perforation rates)
+  getReservoirSolver()->assembleSystem( time_n, dt,
+                                        domain,
+                                        dofManager,
+                                        localMatrix,
+                                        localRhs );
+
+  // assemble J_WW (excluding perforation rates)
+  getWellSolver()->assembleSystem( time_n, dt,
+                                   domain,
+                                   dofManager,
+                                   localMatrix,
+                                   localRhs );
+
+  // assemble perforation rates in J_WR, J_RW, J_RR and J_WW
+  assembleCouplingTerms( time_n, dt,
+                         domain,
+                         dofManager,
+                         localMatrix,
+                         localRhs );
+}
+
+template< typename RESERVOIR_SOLVER, typename WELL_SOLVER >
+void CoupledReservoirAndWellsBase< RESERVOIR_SOLVER, WELL_SOLVER >::
+solveLinearSystem( DofManager const & dofManager,
+                   ParallelMatrix & matrix,
+                   ParallelVector & rhs,
+                   ParallelVector & solution )
+{
+  GEOSX_MARK_FUNCTION;
+
+  rhs.scale( -1.0 );
+  solution.zero();
+  SolverBase::solveLinearSystem( dofManager, matrix, rhs, solution );
+}
+
+template class CoupledReservoirAndWellsBase< SinglePhaseFVM< SinglePhaseBase >, SinglePhaseWell >;
+template class CoupledReservoirAndWellsBase< SinglePhaseHybridFVM, SinglePhaseWell >;
+template class CoupledReservoirAndWellsBase< CompositionalMultiphaseFVM, CompositionalMultiphaseWell >;
+template class CoupledReservoirAndWellsBase< CompositionalMultiphaseHybridFVM, CompositionalMultiphaseWell >;
```

Comment:
Hmm, I don't particularly like neither having to explicitly instantiate a base class with specific sub-solver types (if someone forgets to add an instantiation, they get tons of linker errors), nor the fact that we're recompiling common code 4 times that doesn't even depend (for the most part) on sub-solver types. Moving everything to a header would only solve the first problem. The second one might require multiple virtual inheritance, which I'm not too keen on using. Need to think about this one. @TotoGaz thoughts?

Reply:
> (if someone forgets to add an instantiation, they get tons of linker errors)

Relying on linker errors to make things work is a bit surprising.
I remember we have this kind of tricks in some `Packing` functions...
I cannot find it but I still remember my emotions when I figured our what was going on... 😢 

> we're recompiling common code 4 times that doesn't even depend (for the most part) on sub-solver types.

Then maybe we can split the implementation so OK with this? (Some technical class inherited privately to reduce visibility can help?).
Could we trick into having each of these `template class...` lines in its own `.cpp` file?

> Moving everything to a header would only solve the first problem.

Is is possible to `static_assert` or `enable_if` something such that we keep the templates in the `.cpp` but error messages would be explicit?

> The second one might require multiple virtual inheritance, which I'm not too keen on using. 

This is new kind of `C++` expertise we surely do not want to put into `GEOSX`.
It's already challenging enough.

> @TotoGaz thoughts?

Never, only actions 💪 💥

Reply:
> The second one might require multiple virtual inheritance, which I'm not too keen on using.

Actually a much simpler solution would be to put all code that does not depend on `SOLVERS` pack types into free functions that would get compiled once in the `.cpp` and called from solver methods. The free functions would take solver object as a parameter, along with domain and everything else needed. This leaves a trivial amount of code in the solver member functions, which can move into the header entirely.

Reply:
Ok thanks, I will do that. I implemented all the other suggestions, so I will push now and work on this one. Thanks!

Reply:
Done, I removed 
```
template class CoupledReservoirAndWellsBase< SinglePhaseFVM< SinglePhaseBase >, SinglePhaseWell >;
template class CoupledReservoirAndWellsBase< SinglePhaseHybridFVM, SinglePhaseWell >;
template class CoupledReservoirAndWellsBase< CompositionalMultiphaseFVM, CompositionalMultiphaseWell >;
template class CoupledReservoirAndWellsBase< CompositionalMultiphaseHybridFVM, CompositionalMultiphaseWell >;
```
in the base class, moved the function implementations in the base class, and added free functions to move (almost all) the implementations details to the cpp. For one function (`setupSystem`), I left the body in the hpp, because the implementations depends on the derived types.